gpytorch/lazy/lazy_tensor.py

```diff
@@ -1329,6 +1329,39 @@ def requires_grad_(self, val):
         self.requires_grad = val
         return self
 
+    @cached(name="diagonalization")
+    def diagonalization(self, method: Optional[str] = None):
+        if not self.is_square:
+            raise RuntimeError(
+                "diagonalization only operates on (batches of) square (symmetric) LazyTensors. "
```

Comment:
I we wanted to be fancy we could also allow diagonalization of non-symmetric matrices if we don't restrict things to unitary matrices. E.g. for a `TriangularLazyTensor` we can easily read off the evals and construct the evecs by backward substitution. I guess that's something for down the road maybe...